design/sync-v2.md

```diff
@@ -0,0 +1,216 @@
+* The last updated date should be identified by a new attribute `[UpdatedAt]` and must be an auto-incrementing comparable.  For example, we can use a `long` or a `DateTimeOffset`.  The only requirement is that it must be every increasing so we can use a `>` comparison.
+* The version must either be specified by a `[Version]` filter (and can either be a string or byte array), or the DTO must implement `IDatasyncVersion` which has a `.GetVersion()` (returns a string) to generate the version.  This allows the developer to decide how to compute a version.
+* A `Deleted` flag may be provided by using `[Deleted]` on a field.
+
+For a `TodoItem`, for example:
+
+```csharp
+class TodoItemDTO : IDatasyncVersion
+{
+    [Key]
+    public long Id { get; set; }
+
+    [UpdatedAt]
+    public DateTimeOffset UpdatedAt { get; set; }
+
+    public string IDatasyncVersion.GetVersion()
+        => UpdatedAt.UtcTicks.ToString();
+
+    public string Title { get; set; }
+
+    public string Description { get; set; }
+
+    public DateOnly DueDate { get; set; }
+
+    public TimeSpan AlertBeforeDueDate { get; set; }
+
+    public TimeOnly AlertTime { get; set; }
+
+    public DateTimeOffset? CompletedDate { get; set; }
+
+    public UIColor StatusColor { get; set; }
+}
+```
+
+You will notice:
+
+* We rely on the database to handle the ID and UpdatedAt.
+* The version is computed from the UpdatedAt field - no need to keep something extra.
+* There are a variety of types here; we should artificially restrict the types.
+* The DTO should be able to be handled on both client and server, potentially mapping using Automapper to get the DatabaseGenerated field for UpdatedAt.
+
+This should also be handled quite easily with non-OData requests; for example, GraphQL, gRPC, and other transports should be easily provided.
+
+## Local Service
+
+On the local side, we should utilize common libraries, like Entity Framework Core, to implement the offline synchronization library.  This is split into two parts - 
+
+1. When the client application accesses the local database
+2. When the synchronization process makes changes to the local database.
+
+Ideally, the same DTO should be able to be used on both sides (remote service and offline database).  
+
+When a client application makes a change to the local database:
+
+1. The change is validated.
+2. The operation is written to the operations queue.
+3. The change is written to the local offline table.
+
+When the synchronization process makes a change to the local database:
+
+1. The change is validated.
+2. The operation is removed from the operations queue.
+3. The change is written to the local offline table.
+
+In both cases, the operations queue and offline table change needs to be written in a transaction so that it may be rolled back in its entirety. 
```

Comment:
Probably out of scope, but throwing it out there with no expectation it would be implemented -- it would be really cool if the operation queue was updated by SQLite database triggers (at least in the case of the SQLite DatasyncClient) -- this would allow other apps to modify the same SQLite database locally and changes would be tracked in the operation queue by the database and not by the app layer. Triggers would probably have to be created during the DefineTable as an async overload/extension to indicate that the operation queue should be trigger based.

Reply:
Given the relative complexity of the operations queue (JSON serialization in the context of the client), I don't think this is possible.